tools/publicapigen/main.go

```diff
@@ -0,0 +1,463 @@
+							{
+								Slash: token.NoPos,
+								Text:  "// for godoc to show unexported fields",
+							},
+						}},
+					})
+				}
+			}
+
+			return true
+		},
+	)
+	return nil
+}
+
+// walkDir recursively descends path, calling walkFn for directory
+func walkDir(dir, rel string, f func(path, rel string, files []os.FileInfo) error) error {
+	if rel == "types/uuid" {
+		// we don't want to rewrite this package
+		return nil
+	}
+
+	log.Debug().Str("rel", rel).Msg("Walking directory")
+	entries, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	// Split the files and dirs
+	var dirs, files []os.FileInfo
+	for _, entry := range entries {
+		if entry.IsDir() {
+			dirs = append(dirs, entry)
+		} else {
+			files = append(files, entry)
+		}
+	}
+
+	if err := f(dir, rel, files); err != nil {
+		return fmt.Errorf("error walking %s: %w", rel, err)
+	}
+
+	for _, d := range dirs {
+		dir2 := filepath.Join(dir, d.Name())
+		rel2 := path.Join(rel, d.Name())
+
+		if err := walkDir(dir2, rel2, f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func repoDir() string {
+	_, file, _, ok := runtime.Caller(0)
+	if !ok {
+		panic("Unable to get caller location")
+	}
+
+	publicapigenDir := filepath.Dir(file)
+	toolsDir := filepath.Dir(publicapigenDir)
+	encoreDir := filepath.Dir(toolsDir)
+
+	return encoreDir
+}
+
+func outDir() string {
+	srcDir := filepath.Dir(repoDir())
+	return filepath.Join(srcDir, "encore.dev")
+}
+
+// typeName returns the identifier of the expression unwrapping any pointers, selectors or generics, or nil if it is not an identifier
+func typeName(node ast.Expr) *ast.Ident {
+	// Remove any wrapped references
+	for {
+		if star, ok := node.(*ast.StarExpr); ok {
+			node = star.X
+			continue
+		}
+
+		if index, ok := node.(*ast.IndexExpr); ok {
+			node = index.X
+			continue
+		}
+
+		if indexList, ok := node.(*ast.IndexListExpr); ok {
+			node = indexList.X
+			continue
+		}
+
+		if selector, ok := node.(*ast.SelectorExpr); ok {
+			node = selector.X
+		}
+
+		break
+	}
+
+	if ident, ok := node.(*ast.Ident); ok {
+		return ident
+	}
+	return nil
+}
+
+func clearCommentGroup(node *ast.CommentGroup) {
+	if node == nil {
+		return
+	}
+	node.List = []*ast.Comment{node.List[0]}
+	node.List[0].Text = "  " // double space to prevent a panic when printing the file back out
+}
+
+func mustKeep(nodes ...*ast.CommentGroup) bool {
+	for _, node := range nodes {
+		if node != nil && node.List != nil {
+			for i, comment := range node.List {
+				if comment.Text == "//encore:keep" {
+					if i == 0 {
+						comment.Text = "  "
+					} else {
+						comment.Text = "//" // empty comment line as I want the docs to remain active, but I can't remove this without causing a blank line between the comment group and what ever it's associated with
+					}
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func isPrivateFile(fileName string) bool {
+	return strings.HasPrefix(fileName, "runtime/") ||
```

Comment:
Are we sure these are always prefixes (and relative paths as a result) and not absolute paths?